benches/barchart.rs

```diff
@@ -0,0 +1,63 @@
+use criterion::{criterion_group, criterion_main, Bencher, BenchmarkId, Criterion};
+use rand::Rng;
+use ratatui::{
+    buffer::Buffer,
+    layout::Rect,
+    widgets::{Bar, BarChart, BarGroup, Widget},
+};
+
+/// Benchmark for rendering a barchart.
+pub fn barchart(c: &mut Criterion) {
+    let mut group = c.benchmark_group("barchart");
+    let mut rng = rand::thread_rng();
+
+    for data_count in [64, 256, 2048] {
+        let data: Vec<Bar> = (0..data_count)
+            .map(|i| {
+                Bar::default()
+                    .label(format!("B{i}").into())
+                    .value(rng.gen_range(0..data_count))
+            })
+            .collect();
+
+        let bargroup = BarGroup::default().bars(&data);
+
+        // Render a basic barchart
+        group.bench_with_input(
+            BenchmarkId::new("render", data_count),
+            &BarChart::default().data(bargroup.clone()),
+            render,
+        );
+
+        // Render a barchart with multiple groups
+        group.bench_with_input(
+            BenchmarkId::new("render_grouped", data_count),
+            &BarChart::default()
+                // We call `data` multiple time to add multiple groups.
+                // This is not a duplicated call.
+                .data(bargroup.clone())
+                .data(bargroup.clone())
+                .data(bargroup.clone()),
+            render,
+        );
+    }
+
+    group.finish();
+}
+
+/// Render the widget in a classical size buffer
+fn render(bencher: &mut Bencher, barchart: &BarChart) {
+    let mut buffer = Buffer::empty(Rect::new(0, 0, 200, 50));
```

Comment:
I am not sure if the measurement of `256` and `2048` bars are accurate. I think maybe we should make the buffer bigger. (for 2048 bars I would reserve at least `2048 * 2 = 4096`, because of the `bar_gap`. maybe a little bit more)

Reply:
Most of the value we get out of the benchmarks is when they find things that are hideously excessively inefficient in normal usage, and so we test against screen sizes of similar amounts, with 200x50 being a readable size for a 16in laptop. I'd imagine that reasonable resolution goes up significantly on wide screens / 8K screens though, so perhaps it's worth picking a standard testing resolution that larger screens into account. Thoughts?

We also currently have a hard limit on cells of 65535 (i.e. u16::MAX - I'm unsure what the original limitation for this constraint was, but it's perhaps worth revisiting).

Reply:
true. It is just that I misunderstood the result as: how much it takes to render `64`, `246` and `2048` bars.

Reply:
> I am not sure if the measurement of 256 and 2048 bars are accurate.

I think they are, they're just not very explicit. They don't test the render of 2048 bars since there is not enough space in the buffer. It's more of "test the computation of 2048 bars", so the results are quite good, even with large datasets, it's pretty fast to compute and then render a subset.

> 200x50 being a readable size for a 16in laptop

That was my first logic, yes. But actually you're right, we should test this with the maximum available resolution (or at least maximum reasonable resolution). Otherwise we would only bench for a subset of the lib users.

> We also currently have a hard limit on cells of 65535

We'd have to test in a real world scenario I think, render a fullscreen terminal on an 8k display and see if it fits.